scipy/special/xsf/cdflib.h

```diff
@@ -0,0 +1,89 @@
+#pragma once
+
+#include "cephes/igam.h"
+#include "config.h"
+#include "error.h"
+#include "tools.h"
+
+namespace xsf {
+
+XSF_HOST_DEVICE inline double gdtrib(double a, double p, double x) {
+    if (std::isnan(p) || std::isnan(a) || std::isnan(x)) {
+        return std::numeric_limits<double>::quiet_NaN();
+    }
+    if (!((0 <= p) && (p <= 1))) {
+        set_error("gdtrib", SF_ERROR_DOMAIN, "Input parameter p is out of range");
+        return std::numeric_limits<double>::quiet_NaN();
+    }
+    if (!(a > 0)) {
+        set_error("gdtrib", SF_ERROR_DOMAIN, "Input parameter a is out of range");
+        return std::numeric_limits<double>::quiet_NaN();
+    }
+    if (!(x >= 0)) {
+        set_error("gdtrib", SF_ERROR_DOMAIN, "Input parameter x is out of range");
+        return std::numeric_limits<double>::quiet_NaN();
+    }
+    if (x == 0.0) {
+        if (p == 0.0) {
+            set_error("gdtrib", SF_ERROR_DOMAIN, "Indeterminate result for (x, p) == (0, 0).");
+            return std::numeric_limits<double>::quiet_NaN();
+        }
+        /* gdtrib(a, p, x) tends to 0 as x -> 0 when p > 0 */
+        return 0.0;
+    }
+    if (p == 0.0) {
+        /* gdtrib(a, p, x) tends to infinity as p -> 0 from the right when x > 0. */
+        set_error("gdtrib", SF_ERROR_SINGULAR, NULL);
+        return std::numeric_limits<double>::infinity();
+    }
+    if (p == 1.0) {
+        /* gdtrib(a, p, x) tends to 0 as p -> 1.0 from the left when x > 0. */
+        return 0.0;
+    }
+
+    double q = 1.0 - p;
+    double (*func)(double, std::tuple<double, double, double>);
+    std::tuple<double, double, double> args;
+    if (p <= q) {
+        args = {a, p, x};
+        func = +[](double b, std::tuple<double, double, double> args) -> double {
+            auto [a, p, x] = args;
+            return cephes::igam(b, a * x) - p;
+        };
+    } else {
+        args = {a, q, x};
+        func = +[](double b, std::tuple<double, double, double> args) -> double {
+            auto [a, q, x] = args;
+            return q - cephes::igamc(b, a * x);
+        };
+    }
+    double lower_bound = std::numeric_limits<double>::min();
+    double upper_bound = std::numeric_limits<double>::max();
+    auto [xl, xr, f_xl, f_xr, bracket_status] =
+        detail::bracket_root_for_cdf_inversion(func, 0.1, 1.0, lower_bound, upper_bound, 10.0, false, 100, args);
```

Comment:
A natural question from a curious reader: why $[0.1,1]$ as the initial bracket? I guess it’s good enough to “kick off” the bracket finder for most practical purpose. Though I think there might be pathological inputs (with very large `a * x` and moderate `p`) where even increasing the bracket by 10 for 100 times (the result is still much less than the max float64) may not be enough. One possible fix for such case could be to make the bracket finder check `boundary` after it runs out of iterations and before it returns.

Reply:
I have `maxiter` set low here for debugging purposes and forgot to set it back. It should be large enough that the search can get all the way to `DBL_MAX`. `maxiter` is only there to prevent a bug from causing an infinite loop.

I'm still playing around with initial values. I think the general idea though is that we don't want the bracket size to increase in a relative sense (in terms of how many unique floating point numbers are in it). What I have works fine in this sense when the bracket moves to the right. If we start if `[1, 10]` or with `[0.1, 1]` and use a factor of ten, then the bracket goes from `[1, 10]` to `[10, 100]`, `[100, 1000`]. 

When the bracket moves to the left we go in one step to `[DBL_MIN, 0.1]`, which means the root finder has a lot of search space. You can see the impact this ends up having below:

```
In [3]: gdtr(2.1, 0.02, 0.2)
Out[3]: np.float64(0.9864949416659149)

In [4]: %timeit gdtrib(2.1, _31, 0.2)
12.1 μs ± 5.42 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

```

compared with

```
In [5]: gdtr(2.1, 1.5, 0.5)
Out[5]: np.float64(0.44808723666028993)

In [6]: %timeit gdtrib(2.1, _35, 0.5)
2.64 μs ± 1.95 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

It seems like it might be a good idea to allow different step strategies when moving the bracket to the left verse to the right.

Reply:
Good point. Since the shape parameter is required to be positive, to make the step size more symmetric might be “just” a matter adjusting the frontier in log scale (or equivalently, reparameterize `b = exp(c)`).

However, I’m not too familiar with the gamma distribution (especially for shape parameter less than one), so not sure if it will work reliably.